### DISCUSSION TITLE
Return body in REGISTER response

### DISCUSSION THREAD

**Discussion by pl-garcia** [fonoster/routr#273] [Q&A]
Hello all,
I want to use routr with HIKvision intercom. The issue I encountered is that the register response must send xml data to the client.
So, I started writing a middleware using the supplied template.
I set-up the content-type header, the content-length header (or not) and the body.
The first two are sent back to the client correctly (trace with tcpdump) but the body is never sent.

Here is my code snippet
```typescript
export const addXMLContent = (request: MessageRequest) =>{
  const req = H.deepCopy(request);
  const buf = Buffer.alloc(16);
  v5("http://routr",v5.URL, buf );
  const rndStr = buf.toString('base64');

  req.message.body = `<regXML>
  <version>V2.0.0</version>
  <random>${rndStr}</random>
  </regXML>`;

  req.message.contentLength = {contentLength: req.message.body.length};
  const r = [
    {
      name: 'Content-Type',
      value: 'text/xml'
    },
    {
      name: 'Content-Length',
      value: `${req.message.body.length}`
    }
  ];

  req.message.extensions = [...req.message.extensions, ...r]
  return req;
}

export const registerHandler = async (
  locationClient: LocationClient,
  req: MessageRequest,
  res: Response
) => {

  if (ProcessHelper.isTypeResponse(req)) {
    if(ProcessHelper.isOk(req)) {
      const req2 = addXMLContent(req);
      const ret = res.send(req2);
      return ret;
    } else {
      return res.send(req);
    }
  }
  else {
    return res.send(
      pipe(
        req,
      )
    );
  }
};
```

grpc is pretty hard to debug but I stepped into the code until the message is converted into protobuf and it seems that the body is present if the payload.
I tried to dig into the edge java code. I found that body may be handled (GRPCSipListener.java line 371)
```java
 if (!response.getMessage().getBody().isEmpty()) {
        assert res.getHeader(ContentTypeHeader.NAME) != null;
        res.setContent(response.getMessage().getBody(), (ContentTypeHeader) res.getHeader(ContentTypeHeader.NAME));
}
```
Currently I am stucked and no idea how to solve that.
Any clue ?

Thank you

Pierre-Luc

**Comment by psanders** [comment]
Hi  Pierre-Luc 👋

A few notes and ideas:

- Q: Are there any relevant logs? 
- Note: The Content-Length might be removed as length is [being computed](https://github.com/fonoster/routr/blob/436f01e74a76ff47fa70d1dca3b37bbdaddf32b2/mods/edgeport/src/main/java/io/routr/GRPCSipListener.java#L303) by the EdgePort. 
- Note: I’m clear about the purpose of getting and returning a variable `ret` but seems unnecessary 
- Note: You could remove the pipe since you have a single-value 
- Q: What does sngrep or Wireshark show? Where is the signaling getting failing?

Let me know. I will do my best to help.
